pkg/kritis/review/review.go

```diff
@@ -50,49 +50,130 @@ func New(c *Config) Reviewer {
+	errMsg := "Some images are not attested by all authorities in any GAP policy:\n"
+	for _, image := range badImages {
+		errMsg = errMsg + fmt.Sprintf("- image %s is not attested by\n", image)
+		for gapName, authNames := range missingAttestations[image] {
+			errMsg = errMsg + fmt.Sprintf("  - %+q in gap %s\n", authNames, gapName)
+		}
+	}
+	return errMsg
+}
+
+// ReviewGAP reviews images against generic attestation policies
+// Return error if policies are not well-formed or violations are found.
+func (r Reviewer) ReviewGAP(images []string, gaps []v1beta1.GenericAttestationPolicy, pod *v1.Pod, c metadata.ReadOnlyClient) error {
+	// If no policy found, images should be admitted.
 	if len(gaps) == 0 {
 		glog.Info("No Generic Attestation Policies found")
 		return nil
 	}
 
+	// Check if all GAPs are well-formed.
+	if err := r.checkGAPs(gaps); err != nil {
+		return err
+	}
+
+	images, removed := util.RemoveGloballyAllowedImages(images)
+	glog.Infof("%d images are globally allowed: %+q\n", len(removed), removed)
+
+	images, removed = util.RemoveGapAllowedImages(images, generateGapAllowlist(gaps))
+	glog.Infof("%d images are further gap allowed: %+q\n", len(removed), removed)
+
+	missingAttestations := map[string]map[string][]string{}
+	var badImages []string
 	for _, image := range images {
 		glog.Infof("Check if %s has valid Attestations.", image)
-		imgAttested := false
+		imgAttestedByAnyGap := false
+		missingAttestations[image] = map[string][]string{}
 		for _, gap := range gaps {
-			glog.Infof("Validating against GenericAttestationPolicy %s", gap.Name)
-			// Get all AttestationAuthorities in this policy.
-			auths, err := r.getAttestationAuthoritiesForGAP(gap)
+			isAttested, notAttestedAuthNames, err := r.reviewImageWithGAP(image, gap, c)
 			if err != nil {
 				return err
 			}
-			_, attestedByAny := r.findUnsatisfiedAuths(image, auths, c)
-			if attestedByAny {
-				imgAttested = true
+			if isAttested {
+				imgAttestedByAnyGap = true
 			}
+			missingAttestations[image][gap.Name] = notAttestedAuthNames
 		}
-		if err := r.config.Strategy.HandleAttestation(image, pod, imgAttested); err != nil {
+		if err := r.config.Strategy.HandleAttestation(image, pod, imgAttestedByAnyGap); err != nil {
 			glog.Errorf("error handling attestations %v", err)
 		}
-		if !imgAttested {
-			return fmt.Errorf("image %s is not attested", image)
+		if !imgAttestedByAnyGap {
+			glog.Infof("Image %s is not attested by any gap policy.", image)
+			badImages = append(badImages, image)
 		}
 	}
+	if len(badImages) != 0 {
+		errMsg := generateMissingAttestationSummary(badImages, missingAttestations)
+		return fmt.Errorf(errMsg)
+	}
+
+	if len(images) == 0 {
+		glog.Infof("images are all globally or gap allowed, returning successful status: %s", images)
+		return nil
+	}
+
+	glog.Infof("%d images are further attested: %+q\n", len(images), images)
```

Comment:
What does 'further' imply here? They're attested by something else in addition to this?

Reply:
so here is how the messages will look like:
```
1 images are globally allowed: image-1.
1 images are further gap allowed: image-2.
1 images are further attested: image-3.
```

Reply:
Got it, this is nice formatting! Let's remove 'further', as I think w/o that word it still makes sense.

Reply:
done.